server/src/main/scala/com/cloudera/livy/server/recovery/SessionStore.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to Cloudera, Inc. under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Cloudera, Inc. licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cloudera.livy.server.recovery
+
+import scala.util.{Failure, Success, Try}
+
+import com.cloudera.livy.{LivyConf, Logging}
+import com.cloudera.livy.server.batch.BatchRecoveryMetadata
+import com.cloudera.livy.sessions.Session.RecoveryMetadata
+
+private[recovery] case class SessionManagerState(nextSessionId: Int)
+
+/**
+ * SessionStore provides high level functions to get/save session state from/to StateStore.
+ */
+class SessionStore(
+    livyConf: LivyConf,
+    store: => StateStore = StateStore.get) // For unit testing.
+  extends Logging {
+
+  private val STORE_VERSION: String = "v1"
+  private var nextSessionId = 0
+
+  /**
+   * Persist a session to the session state store.
+   * @param m RecoveryMetadata for the session.
+   */
+  def save(sessionType: String, m: RecoveryMetadata): Unit = synchronized {
```

Comment:
Moved synchronized just to protect nextSessionId.
I don't want to use an atomic int here because the code isn't incrementing the id, but setting it to a max value.